### PR TITLE
Allowing multiple weightings/vehicle using CH

### PR DESCRIPTION
**Pull request by boldtrn**
This PR allows users to setup multiple weightings using CH. You have to define the weightings in the properties file like: 
`prepare.chWeighting=fastest, shortest`

### CODE REVIEW

**Comment by karussell** [comment]
Thanks! This is indeed a very minor change - nice :) !

We could even avoid breaking the Java API when offering the old method setCHWeighting(String).

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 1]
```diff
@@ -297,18 +297,29 @@ private GraphHopper setUnsafeMemory()
...
+     * @param weightingList A list containing multiple weightings like:
+     *                     "fastest", "shortest" or your own weight-calculation type.
      */
-    public GraphHopper setCHWeighting( String weighting )
+    public GraphHopper setCHWeighting( List<String> weightingList )
     {
         ensureNotLoaded();
-        chWeightingStr = weighting.toLowerCase();
+
+        if(weightingList.isEmpty())
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: we should throw an exception here saying at least one weighting is required.

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 2]
```diff
@@ -297,18 +297,29 @@ private GraphHopper setUnsafeMemory()
...
+            return this;
+
+        this.chWeightingList.clear();
+        for (String weight: weightingList)
+        {
+            weight = weight.toLowerCase();
+            weight = weight.trim();
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: should we assume already trimmed strings?

**Comment by karussell** [comment]
Ah, okay the method `public List<String> getCHWeighting()` breaks the Java API and should be added to core/files/changelog.txt

File: config-example.properties [hunk 3]
```diff
@@ -42,6 +42,9 @@ graph.dataaccess=RAM_STORE
 # graph but less for every request.
 # prepare.chWeighting=fastest
 #
+# You can also setup multiple weightings, by providing a coma separated list like:
+# prepare.chWeighting=fastest, shortest
```
Comment (devemux86) [inline config-example.properties]: Probably comma separated without spaces ?

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 4]
```diff
@@ -959,7 +974,8 @@ public Weighting getWeightingForCH( WeightingMap weightingMap, FlagEncoder encod
         // get requested weighting name
         String weightingStr = weightingMap.getWeighting().toLowerCase();
         if (weightingStr.isEmpty())
-            weightingStr = chWeightingStr;
+            // TODO Ugly workaround or valid?
+            weightingStr = getCHWeighting().get(0);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: This is valid as we always assume at least 'fastest' in the weighting list. Maybe throw an exception here if it is empty with a more clear exception than the ArrayOutOfBounds thing

File: config-example.properties (hunk 3 above)
Comment (karussell) [inline config-example.properties]: for the encoders we accept also spaces I think, trimming is easy and should be done before calling setCHWeighting I think. But for now I would be even okay with trimming inside the setter, not sure.

Comment (boldtrn) [inline config-example.properties]: I thought about that as well. What is the downside of trimming? We do not allow weighting names like `" name"` or `"name "`. The benefit is that our config files are not sensible to whitespace.

I would not expect a config file to be sensible to whitespaces.

> trimming inside the setter

Do you think adding logic to a setter is bad?

**Comment by boldtrn** [comment]
> We could even avoid breaking the Java API when offering the old method setCHWeighting(String).

How about creating a setCHWeighting method that uses varargs. This would not break the setter.

> public List<String> getCHWeighting()

How about providing a `String getCHWeighting()` method and add a `@deprecated `. The method would return the first element of list. This method could be also used in the `getWeightingForCH` method. We could remove that method in the future.

Therefore, this would be a pure feature that does not break the api and could be merged soon (since I'd like to see that feature in the master branch ;)).

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (boldtrn) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: You are right. I will update the PR

File: config-example.properties (hunk 3 above)
Comment (karussell) [inline config-example.properties]: > Do you think adding logic to a setter is bad?

Yeah. That is the only thing I don't like about it. But when reading the config file we should trim and then set with the array which is then used 'as it'. Maybe even do not do toLowerCase but throw exception if upper case?

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 5]
```diff
@@ -294,21 +293,57 @@ private GraphHopper setUnsafeMemory()
...
+        for (String weight : weightingList)
+        {
+            weight = weight.toLowerCase();
+            weight = weight.trim();
+            this.chWeightingList.add(weight);
+        }
         return this;
     }
 
+    public List<String> getAllCHWeightings()
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: here I vote for just 'getCHWeightings'

**Comment by karussell** [comment]
Perfect ideas!

Just for the toLowerCase and trimming I would like to know your oppinion e.g. if just doing it or throwing exceptions etc

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 5 above)
Comment (boldtrn) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: Me too, but overloading of methods is only valid if you use different parameters. Same method name with different return type is not valid.

**Comment by boldtrn** [comment]
> Just for the toLowerCase and trimming I would like to know your oppinion e.g. if just doing it or throwing exceptions etc

Actually, I don't think we are doing any harm with the toLowerCase (was there before actually) and the trimming.

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 5 above)
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: Ah, ups, you're right. You wanted to keep the old method, hmmh. I think we should break the API here as the new method should also have a nice name and the old method is probably not that often used.

File: config-example.properties (hunk 3 above)
Comment (boldtrn) [inline config-example.properties]: I actually don't think that adding logic to getters/setter is a bad practice. As long as we are trying to improve the quality of the "seted" value.

The pro of doing it in the setter is that we have a central/unique place to execute the correction of weighting values.

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 5 above)
Comment (boldtrn) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: Ok, so I will change the name from `getAllCHWeightings` to `getCHWeightings` and change the name of `getCHWeighting` to `getFallbackCHWeighting` to be used in the `getWeightingForCH` method.

After writing the above I just realized, that `getWeightingForCH` and `getCHWeighting` is probably irritating.

Maybe we refactor the `getAllCHWeightings ` to `getCHWeightingNames`?

Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: > After writing the above I just realized, that getWeightingForCH and getCHWeighting is probably irritating.

Hmmh, that is right. I would get rid of it and just keep `getCHWeightings` and `getWeightingForCH`. Currently new GH versions broke the Java API and if it is properly documented in the changelog.txt there was no problem or even questions in the forum.

BTW: Maybe we should rename `getWeightingForCH` to `findCHWeighting` to indicate the small overhead? But would be a different issue

**Comment by boldtrn** [comment]
Ok everything should be fixed now. I added the renaming to the changelog.txt. What do you think, can we merge this branch?

**Comment by karussell** [comment]
Thanks! One minor thing now is that the setter and getter of the `List<String>` are slightly 'misnamed'. Would you rename `public GraphHopper setCHWeighting( List<String> weightingList )` to setCHWeighting**s** and then squash your changes to one commit?

**Comment by boldtrn** [comment]
Ok sure, I can do that. But I would leave the fallback method using varargs without the s? To not break the API at this point.

**Comment by karussell** [comment]
> But I would leave the fallback method using varargs without the s? To not break the API at this point.

Hmmh, good question. Or keep setCHWeighting with one string and do setCHWeightings with the varargs?

**Comment by boldtrn** [comment]
> Hmmh, good question. Or keep setCHWeighting with one string and do setCHWeightings with the varargs? As I would mark the old setCHWeighting method as deprecated and remove it at some point (?)

Ok this sounds reasonable. Change both List+Varargs to setCHWeightings and add a setCHWeighting method that is set deprecated?

**Comment by karussell** [comment]
Ok!

**Comment by boldtrn** [comment]
PR is squashed. Thanks.

**Comment by karussell** [comment]
> PR is squashed. Thanks.

Oh, I have to thank: thanks :) !

One thing which I will do in another change:
`String tmpCHWeighting = args.get("prepare.chWeighting", "fastest");`

This should be renamed to `prepare.chWeightings` with a fallback to `prepare.chWeighting` which will be also removed later.

**Comment by karussell** [comment]
Merged with a few changes in the GraphHopper class. https://github.com/graphhopper/graphhopper/commit/641d222d12221aac87bab5a23aefc40fae5edb71

**Comment by karussell** [comment]
Thanks!

**Pull request closed without merging**